linera-core/src/client.rs

```diff
@@ -305,8 +305,36 @@ where
     async fn pending_messages(&mut self) -> Result<Vec<IncomingMessage>, LocalNodeError> {
         let query = ChainInfoQuery::new(self.chain_id).with_pending_messages();
         let response = self.node_client.handle_chain_info_query(query).await?;
+        let mut requested_pending_messages = response.info.requested_pending_messages;
         let mut pending_messages = vec![];
-        for message in response.info.requested_pending_messages {
+        // If this is the first block, it must contain `OpenChain`.
+        if self.next_block_height == BlockHeight::ZERO
+            && self
+                .chain_state_view()
+                .await?
+                .execution_state
+                .system
+                .description
+                .get()
+                .ok_or_else(|| LocalNodeError::InactiveChain(self.chain_id))?
+                .is_child()
+        {
+            let Some(origin) = requested_pending_messages.iter().find_map(|message| {
+                matches!(
+                    message.event.message,
+                    Message::System(SystemMessage::OpenChain { .. })
+                )
+                .then(|| message.origin.clone())
+            }) else {
+                return Err(LocalNodeError::MissingOpenChainMessage);
+            };
+            let (parent_messages, other_messages) = requested_pending_messages
+                .drain(..)
+                .partition(|message| message.origin == origin);
+            requested_pending_messages = other_messages;
+            pending_messages.extend(parent_messages.into_iter().take(self.max_pending_messages));
+        }
+        for message in requested_pending_messages {
             if pending_messages.len() >= self.max_pending_messages {
```

Comment:
Yes, in that case we would _have_ to skip some of those messages.

So we would just enforce that each child chain's first message is `OpenChain`: https://github.com/linera-io/linera-protocol/issues/1227

(I actually had that implemented at first but then realized that the `*_transfer_then_open_chain` tests go exactly against that. We could probably just remove them.)

Reply:
I added it to this PR; it wasn't a big change.

Reply:
We should update the test `*_transfer_then_open_chain` to use two different senders: one for the transfer, then one for open_chain.

I still would like to disallow the `transfer_then_open_chain` when the sender is the same. Even if the transfer can be skipped, it is a bit deceiving to allow sending messages that cannot be possibly received.

Reply:
(Also, we may have messages that are not skippable.)

Reply:
The test validators use `with_allow_inactive_chains(false)`, because we're doing that in production, too. But that means they will reject the transfer messages at first.

If I call `client.receive_certificate(transfer_certificate)`, I can make the client aware of the messages, and propose a block receiving them, but the validators will reject it because they don't have the messages in the inbox.

I fixed it by creating another block on the sender chain _after_ the target chain has been created.